ledger/eval/eval.go

```diff
@@ -1370,43 +1482,153 @@ func (eval *BlockEvaluator) endOfBlock() error {
-		isOnline := acctDelta.Status == basics.Online
-		// true if the accounts last valid round has passed
-		pastCurrentRound := acctDelta.VoteLastValid < currentRound
+		// Regardless of being online or suspended, if voting data exists, the
+		// account can be expired to remove it.  This means an offline account
+		// can be expired (because it was already suspended).
+		if !acctDelta.VoteID.IsEmpty() {
+			expiresBeforeCurrent := acctDelta.VoteLastValid < current
+			if expiresBeforeCurrent &&
+				len(updates.ExpiredParticipationAccounts) < maxExpirations {
+				updates.ExpiredParticipationAccounts = append(
+					updates.ExpiredParticipationAccounts,
+					accountAddr,
+				)
+				continue // if marking expired, do not also suspend
+			}
+		}
+
+		if len(updates.AbsentParticipationAccounts) >= maxSuspensions {
+			continue // no more room (don't break the loop, since we may have more expiries)
+		}
 
-		if isOnline && pastCurrentRound {
-			eval.block.ParticipationUpdates.ExpiredParticipationAccounts = append(
-				eval.block.ParticipationUpdates.ExpiredParticipationAccounts,
-				accountAddr,
-			)
+		if acctDelta.Status == basics.Online {
+			lastSeen := max(acctDelta.LastProposed, acctDelta.LastHeartbeat)
+			if isAbsent(eval.state.prevTotals.Online.Money, acctDelta.MicroAlgos, lastSeen, current) ||
+				failsChallenge(ch, accountAddr, lastSeen) {
+				updates.AbsentParticipationAccounts = append(
+					updates.AbsentParticipationAccounts,
+					accountAddr,
+				)
+			}
 		}
 	}
 }
 
+// delete me in Go 1.21
+func max(a, b basics.Round) basics.Round {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// bitsMatch checks if the first n bits of two byte slices match. Written to
+// work on arbitrary slices, but we expect that n is small. Only user today
+// calls with n=5.
+func bitsMatch(a, b []byte, n int) bool {
+	// Ensure n is a valid number of bits to compare
+	if n < 0 || n > len(a)*8 || n > len(b)*8 {
+		return false
+	}
+
+	// Compare entire bytes when n is bigger than 8
+	for i := 0; i < n/8; i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	// Compare bits in the last byte
+	for i := n / 8 * 8; i < n; i++ {
+		// Calculate the byte and bit positions
+		bytePos := i / 8
+		bitPos := i % 8
+
+		// Extract the bits from each slice
+		bit1 := (a[bytePos] >> (7 - bitPos)) & 1
+		bit2 := (b[bytePos] >> (7 - bitPos)) & 1
+
+		// Compare the bits
+		if bit1 != bit2 {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isAbsent(totalOnlineStake basics.MicroAlgos, acctStake basics.MicroAlgos, lastSeen basics.Round, current basics.Round) bool {
+	// Don't consider accounts that were online when mining went into effect as
+	// absent.  They get noticed the next time they propose or keyreg, which
+	// ought to be soon, if they want to earn incentives.
+	if lastSeen == 0 {
+		return false
+	}
+	// See if the account has exceeded 10x their expected observation interval.
+	allowableLag := basics.Round(10 * totalOnlineStake.Raw / acctStake.Raw)
```

Comment:
Perhaps use `basics.Muldiv` to ensure no funny 64-bit math issues arise?  Shouldn't happen on mainnet due to a relatively small number of total algos, of course.

Reply:
```
	// See if the account has exceeded 10x their expected observation interval.
	allowableLag, o := basics.Muldiv(10, totalOnlineStake.Raw, acctStake.Raw)
	if o {
		// This can't happen with 10B total possible stake, but if we imagine
		// another algorand network with huge possible stake, this seems reasonable.
		allowableLag = math.MaxInt64 / acctStake.Raw
	}
```